crates/ruma-events/src/call/member.rs

```diff
@@ -0,0 +1,386 @@
+    pub fn is_expired(&self, origin_server_ts: Option<Timestamp>) -> bool {
+        let ev_created_ts = match (self.created_ts, origin_server_ts) {
+            (Some(created_ts), Some(_)) => Some(created_ts),
+            (None, Some(server_ts)) => Some(server_ts),
+            (Some(created_ts), None) => Some(created_ts),
+            _ => None,
+        };
+        if let Some(ev_created_ts) = ev_created_ts {
+            let now = Timestamp::now().to_system_time();
+            let expire_ts =
+                ev_created_ts.to_system_time().map(|t| t + Duration::from_millis(self.expires));
+            now > expire_ts
+        } else {
+            // This should not be reached since we only allow events that have copied over
+            // the origin server ts. `copy_origin_server_ts_to_membership`
+            warn!("Encountered a Call Member state event where the origin_ts (or origin_server_ts) could not be found.
+            It is treated as a non expired event but this might be wrong.");
+            false
+        }
+    }
+}
+
+/// Description of the SFU/Foci a membership can be connected to.
+/// Foci is singular for Focus a focus can be any server powering the matrixRTC session (SFU, MCU).
+/// It serves as a node to redistribute rtc streams.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[serde(tag = "type")]
+#[serde(rename_all = "snake_case")]
+#[non_exhaustive]
+pub enum Foci {
+    /// Livekit is one possible type of SFU/Foci that can be used for a matrixRTC session
+    Livekit(LivekitFoci),
+}
+
+/// The fields to describe livekit as an `active_foci`
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[non_exhaustive]
+pub struct LivekitFoci {
+    /// The alias where the livekit sessions can be reached
+    #[serde(rename = "livekit_alias")]
+    pub alias: String,
+    /// The url of the jwt server of the used livekit instance
+    #[serde(rename = "livekit_service_url")]
+    pub service_url: String,
+}
+
+/// The type of the matrixRTC session.
+/// (this is not the application/client used by the user but the
+/// type of matrixRTC session e.g. calling, third-room, whiteboard could be
+/// possible applications.)
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[serde(tag = "application")]
+#[non_exhaustive]
+pub enum Application {
+    #[serde(rename = "m.call")]
+    /// A VoIP call
+    Call(CallApplicationContent),
+    /// Any other application that is not Specced
+    Unknown(serde_json::Value),
+}
+
+/// Call specific parameters membership parameters.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[non_exhaustive]
+pub struct CallApplicationContent {
+    /// An optional identifier for calls. Only relevant for some calls.
+    pub call_id: String,
+    /// Who owns/joins/controls (can modify) the call.
+    pub scope: CallScope,
+}
+
+/// The call scope defines different call ownership models.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
```

Comment:
Here you should use the `StringEnum` derive macro, it will allow to support easily unknown variants, and to rename all variants with the `m.*` prefix.

Reply:
So I can remove the renames and replace it with a `#[serde(rename_all = "snake_case")]`

Reply:
Actually since it's a custom macro the syntax is slightly different. You can copy it from: https://github.com/ruma/ruma/blob/f4c27bb64168940fdb628211caf730fcfcaf4d46/crates/ruma-events/src/location.rs#L180-L196